pkg/webhook/handlers.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// Defaultor defines functions for setting defaults on resources
+type Defaulter interface {
+	runtime.Object
+	Default()
+}
+
+// Validator defines functions for validating an operation
+type Validator interface {
+	runtime.Object
+	ValidateCreate() error
+	ValidateUpdate(old runtime.Object) error
+}
+
+// newMutatingWebhookFor creates a new WebhookBuilder for Validation and Defaulting the provided type.  Returns nil
+// if o does not implement either validator or defaulter.
+func newMutatingWebhookFor(o runtime.Object) *admission.Webhook {
+	d, ok := o.(Defaulter)
+	if !ok {
+		// Type doesn't implement function for mutating webhook
+		return nil
+	}
+	return &admission.Webhook{
+		Handler: &mutatingHandler{defaulter: d},
+	}
+}
+
+type mutatingHandler struct {
+	defaulter Defaulter
+	scheme    *runtime.Scheme
+	decoder   *admission.Decoder
+}
+
+var _ inject.Scheme = &mutatingHandler{}
+
+// InjectDecoder injects the decoder into the FirstMateDeleteHandler
+func (h *mutatingHandler) InjectScheme(s *runtime.Scheme) error {
+	h.scheme = s
+	return nil
+}
+
+//var _ admission.DecoderInjector = &mutatingHandler{}
+//
+//// InjectDecoder injects the decoder into the FirstMateDeleteHandler
+//func (h *mutatingHandler) InjectDecoder(d *admission.Decoder) error {
+//	h.decoder = d
+//	return nil
+//}
+
+// Handle handles admission requests.
+func (h *mutatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	var err error
+	h.decoder, err = admission.NewDecoder(h.scheme)
+	if err != nil {
+		log.Error(err, "unable to get a decoder from the scheme")
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
```

Comment:
I am sure we must have explored that but wanted to understand why this can't be done `once` for this handler at the start-up or may be at the time of scheme injection ?